File: 2026-1/sesion-de-clase-10/usa-procedimientos-almacenados/SoftProg/SoftProgPersistencia/src/main/java/pe/edu/pucp/softprog/dao/ventas/OrdenVentaDAOImpl.java
package pe.edu.pucp.softprog.dao.ventas;

import pe.edu.pucp.softprog.dao.DefaultBaseDAO;
import pe.edu.pucp.softprog.dao.almacen.ProductoDAOImpl;
import pe.edu.pucp.softprog.dao.clientes.ClienteDAOImpl;
import pe.edu.pucp.softprog.dao.rrhh.EmpleadoDAOImpl;
import pe.edu.pucp.softprog.modelo.ventas.LineaOrdenVenta;
import pe.edu.pucp.softprog.modelo.ventas.OrdenVenta;

import java.sql.Connection;
import java.sql.CallableStatement;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Types;
import java.util.ArrayList;
import java.util.List;

public class OrdenVentaDAOImpl extends DefaultBaseDAO<OrdenVenta> implements OrdenVentaDAO {
	@Override
	public Integer crear(OrdenVenta modelo) {
		return ejecutarComando(conn -> {
			Integer idOrden = this.ejecutarComandoCrear(conn, modelo);
			if (idOrden == null) {
				return null;
			}
			modelo.setId(idOrden);
			this.crearLineas(conn, idOrden, modelo.getLineas());
			return idOrden;
		});
	}

	@Override
	public boolean actualizar(OrdenVenta modelo) {
		return ejecutarComando(conn -> {
			if (!this.ejecutarComandoActualizar(conn, modelo)) {
				return false;
			}

			this.eliminarLineasPorOrden(conn, modelo.getId());

			this.crearLineas(conn, modelo.getId(), modelo.getLineas());
			return true;
		});
	}

	@Override
 	public boolean eliminar(Integer id) {
		return ejecutarComando(conn -> {
			this.eliminarLineasPorOrden(conn, id);
			return this.ejecutarComandoEliminar(conn, id);
		});
	}

	@Override
	public OrdenVenta leer(Integer id) {
		return ejecutarComando(conn -> {
			try (PreparedStatement cmd = this.comandoLeer(conn, id);
				 ResultSet rs = cmd.executeQuery()) {
				if (!rs.next()) {
					System.err.println("No se encontro el registro con id: " + id);
					return null;
				}

				OrdenVenta modelo = this.mapearModelo(rs);
				modelo.setLineas(this.leerLineas(conn, modelo.getId()));
				return modelo;
			}
		});
	}

	@Override
	public List<OrdenVenta> leerTodos() {
		return ejecutarComando(conn -> {
			try (PreparedStatement cmd = this.comandoLeerTodos(conn);
				 ResultSet rs = cmd.executeQuery()) {
				List<OrdenVenta> modelos = new ArrayList<>();
				while (rs.next()) {
					OrdenVenta modelo = this.mapearModelo(rs);
					modelo.setLineas(this.leerLineas(conn, modelo.getId()));
					modelos.add(modelo);
				}
				return modelos;
			}
		});
	}

	protected PreparedStatement comandoCrear(Connection conn,
											 OrdenVenta modelo) throws SQLException {
		String sql = "{call insertarOrdenVenta(?, ?, ?, ?, ?)}";
		CallableStatement cmd = conn.prepareCall(sql);
		cmd.setInt("p_idCliente", modelo.getCliente().getId());
		if (modelo.getEmpleado() == null) {
			cmd.setNull("p_idEmpleado", Types.INTEGER);
		}
		else {
			cmd.setInt("p_idEmpleado", modelo.getEmpleado().getId());
		}
		cmd.setDouble("p_total", modelo.getTotal());
		cmd.setBoolean("p_activo", modelo.isActivo());
		cmd.registerOutParameter("p_id", Types.INTEGER);
		return cmd;
	}

	protected PreparedStatement comandoActualizar(Connection conn,
												  OrdenVenta modelo) throws SQLException {
		String sql = "{call modificarOrdenVenta(?, ?, ?, ?, ?)}";
		CallableStatement cmd = conn.prepareCall(sql);
		cmd.setInt("p_idCliente", modelo.getCliente().getId());
		if (modelo.getEmpleado() == null) {
			cmd.setNull("p_idEmpleado", Types.INTEGER);
		}
		else {
			cmd.setInt("p_idEmpleado", modelo.getEmpleado().getId());
		}
		cmd.setDouble("p_total", modelo.getTotal());
		cmd.setBoolean("p_activo", modelo.isActivo());
		cmd.setInt("p_id", modelo.getId());
		return cmd;
	}

	protected PreparedStatement comandoEliminar(Connection conn,
												Integer id) throws SQLException {
		String sql = "{call eliminarOrdenVenta(?)}";
		CallableStatement cmd = conn.prepareCall(sql);
		cmd.setInt("p_id", id);
		return cmd;
	}

	protected PreparedStatement comandoLeer(Connection conn,
											Integer id) throws SQLException {
		String sql = "{call buscarOrdenVentaPorId(?)}";
		CallableStatement cmd = conn.prepareCall(sql);
		cmd.setInt("p_id", id);
		return cmd;
	}

	protected PreparedStatement comandoLeerTodos(Connection conn) throws SQLException {
		String sql = "{call listarOrdenesVenta()}";
		return conn.prepareCall(sql);
	}

	protected PreparedStatement comandoCrearLinea(Connection conn,
                                                  Integer idOrdenVenta,
                                                  LineaOrdenVenta linea) throws SQLException {
		String sql = "{call insertarLineaOrdenVenta(?, ?, ?, ?, ?, ?)}";
		CallableStatement cmd = conn.prepareCall(sql);
		cmd.setInt("p_idOrdenVenta", idOrdenVenta);
		cmd.setInt("p_idProducto", linea.getProducto().getId());
		cmd.setInt("p_cantidad", linea.getCantidad());
		cmd.setDouble("p_subTotal", linea.getSubTotal());
		cmd.setBoolean("p_activo", linea.isActivo());
		cmd.registerOutParameter("p_id", Types.INTEGER);
		return cmd;
	}

	protected PreparedStatement comandoLeerLineas(Connection conn,
                                                  Integer idOrdenVenta) throws SQLException {
		String sql = "{call listarLineasPorOrdenVenta(?)}";
		CallableStatement cmd = conn.prepareCall(sql);
		cmd.setInt("p_idOrdenVenta", idOrdenVenta);
		return cmd;
	}

	protected PreparedStatement comandoEliminarLinea(Connection conn,
													 Integer idLineaOrdenVenta) throws SQLException {
		String sql = "{call eliminarLineaOrdenVenta(?)}";
		CallableStatement cmd = conn.prepareCall(sql);
		cmd.setInt("p_id", idLineaOrdenVenta);
		return cmd;
	}

	@Override
	protected OrdenVenta mapearModelo(ResultSet rs) throws SQLException {
		OrdenVenta modelo = new OrdenVenta();
		modelo.setId(rs.getInt("id"));
		modelo.setCliente(new ClienteDAOImpl().leer(rs.getInt("idCliente")));

		int idEmpleado = rs.getInt("idEmpleado");
		if (!rs.wasNull()) {
			modelo.setEmpleado(new EmpleadoDAOImpl().leer(idEmpleado));
		}

		modelo.setTotal(rs.getDouble("total"));
		modelo.setActivo(rs.getBoolean("activo"));
		return modelo;
	}

	protected LineaOrdenVenta mapearLinea(ResultSet rs) throws SQLException {
		LineaOrdenVenta linea = new LineaOrdenVenta();
		linea.setId(rs.getInt("id"));
		linea.setProducto(new ProductoDAOImpl().leer(rs.getInt("idProducto")));

		linea.setCantidad(rs.getInt("cantidad"));
		linea.setSubTotal(rs.getDouble("subTotal"));
		linea.setActivo(rs.getBoolean("activo"));
		return linea;
	}

	protected void crearLineas(Connection conn,
                               Integer idOrdenVenta,
                               List<LineaOrdenVenta> lineas) throws SQLException {
		if (lineas == null || lineas.isEmpty()) {
			return;
		}

		for (LineaOrdenVenta linea : lineas) {
			try (PreparedStatement cmd = this.comandoCrearLinea(conn, idOrdenVenta, linea)) {
				if (cmd.executeUpdate() == 0) {
					throw new SQLException("No se pudo insertar una linea de orden de venta");
				}
			}
		}
	}

	protected List<LineaOrdenVenta> leerLineas(Connection conn,
                                               Integer idOrdenVenta) throws SQLException {
		try (PreparedStatement cmd = this.comandoLeerLineas(conn, idOrdenVenta);
			 ResultSet rs = cmd.executeQuery()) {
			List<LineaOrdenVenta> lineas = new ArrayList<>();
			while (rs.next()) {
				lineas.add(this.mapearLinea(rs));
			}
			return lineas;
		}
	}

	protected void eliminarLineasPorOrden(Connection conn,
										  Integer idOrdenVenta) throws SQLException {
		try (PreparedStatement cmdLeerLineas = this.comandoLeerLineas(conn, idOrdenVenta);
			 ResultSet rs = cmdLeerLineas.executeQuery()) {
			while (rs.next()) {
				int idLinea = rs.getInt("id");
				try (PreparedStatement cmdEliminarLinea = this.comandoEliminarLinea(conn, idLinea)) {
					cmdEliminarLinea.executeUpdate();
				}
			}
		}
	}
}

File: 2026-1/sesion-de-clase-12/con-transacciones-sql/SoftProgPersistencia/Dao/Ventas/OrdenVentaDaoImpl.cs
using System.Data;
using System.Data.Common;
using SoftProgModelo.Modelos.Ventas;
using SoftProgModelo.Modelos.Rrhh;
using SoftProgPersistencia.Dao.Almacen;
using SoftProgPersistencia.Dao.Clientes;
using SoftProgPersistencia.Dao.Rrhh;

namespace SoftProgPersistencia.Dao.Ventas;

public class OrdenVentaDaoImpl : DefaultBaseDao<OrdenVenta>, IOrdenVentaDao
{
    public override int Crear(OrdenVenta modelo)
    {
        return EjecutarComando(conn =>
        {
            var idOrden = EjecutarComandoCrear(conn, modelo);
            if (idOrden <= 0)
            {
                return 0;
            }

            CrearLineas(conn, idOrden, modelo.Lineas);
            return idOrden;
        });
    }

    public override bool Actualizar(OrdenVenta modelo)
    {
        return EjecutarComando(conn =>
        {
            if (!EjecutarComandoActualizar(conn, modelo))
            {
                return false;
            }

            using (var cmdEliminar = ComandoEliminarLineas(conn, modelo.Id))
            {
                AdjuntarTransaccionActiva(cmdEliminar);
                cmdEliminar.ExecuteNonQuery();
            }

            CrearLineas(conn, modelo.Id, modelo.Lineas);
            return true;
        });
    }

    public override bool Eliminar(int id)
    {
        return EjecutarComando(conn =>
        {
            using (var cmdEliminar = ComandoEliminarLineas(conn, id))
            {
                AdjuntarTransaccionActiva(cmdEliminar);
                cmdEliminar.ExecuteNonQuery();
            }

            return EjecutarComandoEliminar(conn, id);
        });
    }

    public override OrdenVenta? Leer(int id)
    {
        var orden = new OrdenVenta();
        return EjecutarComando(conn =>
        {
            var existe = false;
            var idOrden = 0;
            var idCliente = 0;
            int? idEmpleado = null;
            var total = 0d;
            var activo = false;

            using (var cmd = ComandoLeer(conn, id))
            {
                AdjuntarTransaccionActiva(cmd);
                using (var reader = cmd.ExecuteReader())
                {
                    if (reader.Read())
                    {
                        existe = true;
                        idOrden = LeerEntero(reader, "id");
                        idCliente = LeerEntero(reader, "idCliente");
                        idEmpleado = LeerEnteroNullable(reader, "idEmpleado");
                        total = LeerDecimal(reader, "total");
                        activo = LeerBooleano(reader, "activo");
                    }
                }

                if (!existe)
                {
                    return null;
                }

                orden = new OrdenVenta
                {
                    Id = idOrden,
                    Total = total,
                    Activo = activo,
                    Cliente = new ClienteDaoImpl().Leer(idCliente)
                        ?? throw new InvalidOperationException($"No existe el cliente con id {idCliente}")
                };

                if (idEmpleado.HasValue)
                {
                    orden.Empleado = new EmpleadoDaoImpl().Leer(idEmpleado.Value);
                }
            }

            orden.Lineas = LeerLineas(conn, idOrden);
            return orden;
        });
    }

    public override List<OrdenVenta> LeerTodos()
    {
        return EjecutarComando(conn =>
        {
            var filas = new List<(int IdOrden, int IdCliente, int? IdEmpleado, double Total, bool Activo)>();
            using var cmd = ComandoLeerTodos(conn);
            AdjuntarTransaccionActiva(cmd);
            using (var reader = cmd.ExecuteReader())
            {
                while (reader.Read())
                {
                    filas.Add((
                        LeerEntero(reader, "id"),
                        LeerEntero(reader, "idCliente"),
                        LeerEnteroNullable(reader, "idEmpleado"),
                        LeerDecimal(reader, "total"),
                        LeerBooleano(reader, "activo")));
                }
            }

            var ordenes = new List<OrdenVenta>(filas.Count);
            foreach (var (IdOrden, IdCliente, IdEmpleado, Total, Activo) in filas)
            {
                var cliente = new ClienteDaoImpl().Leer(IdCliente)
                    ?? throw new InvalidOperationException($"No existe el cliente con id {IdCliente}");

                Empleado? empleado = null;
                if (IdEmpleado.HasValue)
                {
                    empleado = new EmpleadoDaoImpl().Leer(IdEmpleado.Value);
                }

                var orden = new OrdenVenta
                {
                    Id = IdOrden,
                    Total = Total,
                    Activo = Activo,
                    Cliente = cliente,
                    Empleado = empleado,
                    Lineas = LeerLineas(conn, IdOrden)
                };

                ordenes.Add(orden);
            }

            return ordenes;
        });
    }

    protected override DbCommand ComandoCrear(DbConnection conn, OrdenVenta modelo)
    {
        var cmd = conn.CreateCommand();
        cmd.CommandText = "INSERT INTO ORDENVENTA (idCliente, idEmpleado, total, activo) VALUES (@p1, @p2, @p3, @p4);";
        CrearParametro(cmd, "@p1", modelo.Cliente?.Id ?? throw new InvalidOperationException("Cliente es obligatorio"));
        CrearParametro(cmd, "@p2", modelo.Empleado?.Id, DbType.Int32);
        CrearParametro(cmd, "@p3", modelo.Total);
        CrearParametro(cmd, "@p4", modelo.Activo);
        return cmd;
    }

    protected override DbCommand ComandoActualizar(DbConnection conn, OrdenVenta modelo)
    {
        var cmd = conn.CreateCommand();
        cmd.CommandText = @"UPDATE ORDENVENTA SET idCliente = @p1, idEmpleado = @p2, total = @p3, activo = @p4
                            WHERE id = @p5;";
        CrearParametro(cmd, "@p1", modelo.Cliente?.Id ?? throw new InvalidOperationException("Cliente es obligatorio"));
        CrearParametro(cmd, "@p2", modelo.Empleado?.Id, DbType.Int32);
        CrearParametro(cmd, "@p3", modelo.Total);
        CrearParametro(cmd, "@p4", modelo.Activo);
        CrearParametro(cmd, "@p5", modelo.Id);
        return cmd;
    }

    protected override DbCommand ComandoEliminar(DbConnection conn, int id)
    {
        var cmd = conn.CreateCommand();
        cmd.CommandText = "DELETE FROM ORDENVENTA WHERE id = @p1;";
        CrearParametro(cmd, "@p1", id);
        return cmd;
    }

    protected override DbCommand ComandoLeer(DbConnection conn, int id)
    {
        var cmd = conn.CreateCommand();
        cmd.CommandText = "SELECT * FROM ORDENVENTA WHERE id = @p1;";
        CrearParametro(cmd, "@p1", id);
        return cmd;
    }

    protected override DbCommand ComandoLeerTodos(DbConnection conn)
    {
        var cmd = conn.CreateCommand();
        cmd.CommandText = "SELECT * FROM ORDENVENTA;";
        return cmd;
    }

    private DbCommand ComandoCrearLinea(DbConnection conn, int idOrdenVenta, LineaOrdenVenta linea)
    {
        var cmd = conn.CreateCommand();
        cmd.CommandText = @"INSERT INTO LINEAORDENVENTA (idOrdenVenta, idProducto, cantidad, subTotal, activo)
                            VALUES (@p1, @p2, @p3, @p4, @p5);";
        CrearParametro(cmd, "@p1", idOrdenVenta);
        CrearParametro(cmd, "@p2", linea.Producto?.Id ?? throw new InvalidOperationException("Producto es obligatorio"));
        CrearParametro(cmd, "@p3", linea.Cantidad);
        CrearParametro(cmd, "@p4", linea.SubTotal);
        CrearParametro(cmd, "@p5", linea.Activo);
        return cmd;
    }

    private DbCommand ComandoLeerLineas(DbConnection conn, int idOrdenVenta)
    {
        var cmd = conn.CreateCommand();
        cmd.CommandText = "SELECT * FROM LINEAORDENVENTA WHERE idOrdenVenta = @p1;";
        CrearParametro(cmd, "@p1", idOrdenVenta);
        return cmd;
    }

    private DbCommand ComandoEliminarLineas(DbConnection conn, int idOrdenVenta)
    {
        var cmd = conn.CreateCommand();
        cmd.CommandText = "DELETE FROM LINEAORDENVENTA WHERE idOrdenVenta = @p1;";
        CrearParametro(cmd, "@p1", idOrdenVenta);
        return cmd;
    }

    protected override OrdenVenta MapearModelo(DbDataReader reader)
    {
        var orden = new OrdenVenta
        {
            Id = LeerEntero(reader, "id"),
            Total = LeerDecimal(reader, "total"),
            Activo = LeerBooleano(reader, "activo"),
            Cliente = new ClienteDaoImpl().Leer(LeerEntero(reader, "idCliente"))
        };

        var idEmpleado = LeerEnteroNullable(reader, "idEmpleado");
        if (idEmpleado.HasValue)
        {
            orden.Empleado = new EmpleadoDaoImpl().Leer(idEmpleado.Value);
        }

        return orden;
    }

    private void CrearLineas(DbConnection conn, int idOrdenVenta, List<LineaOrdenVenta> lineas)
    {
        if (lineas.Count == 0)
        {
            return;
        }

        foreach (var linea in lineas)
        {
            using var cmd = ComandoCrearLinea(conn, idOrdenVenta, linea);
            AdjuntarTransaccionActiva(cmd);
            _ = cmd.ExecuteNonQuery();
        }
    }

    private List<LineaOrdenVenta> LeerLineas(DbConnection conn, int idOrdenVenta)
    {
        var filas = new List<(int Id, int IdProducto, int Cantidad, double SubTotal, bool Activo)>();
        using (var cmd = ComandoLeerLineas(conn, idOrdenVenta))
        {
            AdjuntarTransaccionActiva(cmd);
            using var reader = cmd.ExecuteReader();
            while (reader.Read())
            {
                filas.Add((
                    LeerEntero(reader, "id"),
                    LeerEntero(reader, "idProducto"),
                    LeerEntero(reader, "cantidad"),
                    LeerDecimal(reader, "subTotal"),
                    LeerBooleano(reader, "activo")));
            }
        }

        var lineas = new List<LineaOrdenVenta>(filas.Count);
        foreach (var (Id, IdProducto, Cantidad, SubTotal, Activo) in filas)
        {
            lineas.Add(new LineaOrdenVenta
            {
                Id = Id,
                Producto = new ProductoDaoImpl().Leer(IdProducto),
                Cantidad = Cantidad,
                SubTotal = SubTotal,
                Activo = Activo
            });
        }

        return lineas;
    }
}
